### PR TITLE
Integrating butterfly to the supervisor

### PR DESCRIPTION
**Pull request by adamhjk**
gif-keyboard-10120060518376514766

This is an incomplete and substantial refactor of the supervisor. It:

* Uses Butterfly for the gossip layer
* Has a `Manager`, that watches multiple `Services`
* `Services` are responsible for their own `Configuration`,
  and will supervise their child processes.
* Has butterfly's test suite broken in to multiple components

Several things remain:

* Warnings need to be cleaned up
* The sidecar needs to be rebuilt and re-enabled
* The `hab` commands that integrate with the ring are non-functional,
  and neccessiate the addition of a `Client` to butterfly.
* The `initializer` topology is non-functional.
* No dead code has been pruned from the tree; it's useful as a reference
  during the refactor, but should absolutely be removed when we are
  complete.

Signed-off-by: Adam Jacob <adam@chef.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/sup/src/config.rs [hunk 1]
```diff
@@ -27,8 +27,6 @@ use std::mem;
 use hcore::package::PackageIdent;
 
 use error::{Error, SupError};
-use gossip::server::GOSSIP_DEFAULT_PORT;
-use topology::Topology;
```
Comment (smurawski) [inline components/sup/src/config.rs]: Was the topology import removal here intentional? (Seems to be the source of the build failure in travis).

File: components/sup/src/config.rs [hunk 2]
```diff
@@ -407,7 +438,7 @@ impl Config {
     pub fn set_gossip_peer(&mut self, mut gp: Vec<String>) -> &mut Config {
         for p in gp.iter_mut() {
             if p.find(':').is_none() {
-                p.push_str(&format!(":{}", GOSSIP_DEFAULT_PORT));
+                p.push_str(&format!(":{}", 9638));
```
Comment (reset) [inline components/sup/src/config.rs]: I think we should leave this assigned to a const

File: components/butterfly/src/server/inbound.rs [hunk 3]
```diff
@@ -109,8 +109,8 @@ impl<'a> Inbound<'a> {
                 }
                 Err(e) => {
                     match e.raw_os_error() {
-                        Some(35) => {
-                            // This is the normal non-blocking result
+                        Some(35) | Some(11) => {
```
Comment (adamhjk) [inline components/butterfly/src/server/inbound.rs]: Good times. We'll certainly have to fix this up - they shouldn't be different between Linux and Mac, given that it's POSIX socket results bubbled up.

Comment (adamhjk) [inline components/butterfly/src/server/inbound.rs]: So, as @reset points out, they are super duper different. The names are POSIX standards; the numbers, well, those can be whatever. :)

**Comment by adamhjk** [comment]
@thesentinels force

**Comment by thesentinels** [comment]
Excellent @adamhjk! It always makes me feel nice when humans approve of one anothers work, no matter the consequences. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**